---
title: "RNAseq_protoplast_DEanalysis_ver2"
author: "Rita Tam"
date: "28/02/2022"
output: html_document
---

### DESeq2 citation
Love MI, Huber W, Anders S (2014). “Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2.” Genome Biology, 15, 550. doi: 10.1186/s13059-014-0550-8.

## Workspace set-up

```{r message=FALSE, warning=FALSE, results="hide"}
library(DESeq2)
library(tximport)
library(tidyverse)
```

Setting up functions for later use.

```{r message=FALSE, warning=FALSE}
create_dir <- function(dirname) {
  if (file.exists(dirname)) {
    cat("This directory already exists\n")
  } else {
    dir.create(dirname)
  }
}
```

Set working directory and input file paths.

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "DE_analysis")
samp_rep_dirs <- list.files(path="input/kallisto_quantification/", full.names=TRUE)
```


# A. Gabo and GaboSr50

## 1. Input

### Transcript abundance files

Specify paths to the abundance files tagged by a column of sample names, and read in a tx2gene table that links transcripts to genes for this dataset.

```{r message=FALSE}
# get absolute paths to all abundance files
abundance_files <- c()
for (path in samp_rep_dirs){
  path <- paste(path,"/abundance.h5", sep="")
  abundance_files <- c(abundance_files, path)
}

samp_rep_names <- list.files(path="input/kallisto_quantification/", full.names=FALSE)
names(abundance_files) <- samp_rep_names

all(file.exists(abundance_files)) # check all files exist
head(abundance_files)

# read in tx2gene file
t2g <- read_delim("input/transcriptome/Triticum_aestivum.cdna.tx2gene.txt", delim=" ", col_names=FALSE)
head(t2g)
```

Slice abundance file list to include only Gabo and GaboSr50.

```{r}
abundance_files_AllGabo <- abundance_files[17:36]
```

### *Tximport*

Using tximport function to import quantification data.

```{r message=FALSE}
txi_AllGabo <- tximport(abundance_files_AllGabo, type="kallisto", tx2gene=t2g)
```

Construct a sample table with condition column.

```{r}
samples_AllGabo <- data.frame(condition=factor(rep(c("Gabo_AvrSr50",
                                             "Gabo_EV",
                                             "Gabo_Sr50_AvrSr35",
                                             "Gabo_Sr50_AvrSr50",
                                             "Gabo_Sr50_EV"),
                                           each=4)))
rownames(samples_AllGabo) <- colnames(txi_AllGabo$counts)
samples_AllGabo
```

### Construct a *DESeqDataSet*

Now with the txi object and sample table, construct a DESeqDataSet.

```{r message=FALSE}
dds_AllGabo <- DESeqDataSetFromTximport(txi_AllGabo, colData=samples_AllGabo, design=~condition)
dds_AllGabo
```

Remove genes with zero reads, to keep only rows that have at least \>1 read in total across all genes.

```{r}
dds_AllGabo <- dds_AllGabo[rowSums(counts(dds_AllGabo)) > 1, ]
dds_AllGabo
```

## 2. Differential expression analysis

The standard differential expression analysis steps are wrapped into a single function, *DESeq()*. Results tables are generated using the function *results()*, which extract log2 fold changes, p values and adjusted p values (for multiple tests).

```{r message=FALSE}
dds_AllGabo <- DESeq(dds_AllGabo)
```

```{r}
dds_AllGabo
```

## 3. Result tables

Applying *summary()* to a result table will return:\
\* LFC \> 0 (up) --> upregulated genes\
\* LFC \< 0 (down) --> downregulated genes\
\* outliers\
\* low counts\

------------------------------------------------------------------------

Setting up *DEanalysis_results()* function that takes in: 1. dds (DESeqDataSet) 2. query condition 3. reference condition

```{r}
DEanalysis_results <- function(dds, query, ref) {
  DEres <- results(dds, contrast=c("condition", query, ref), alpha=0.05)
  DEres <- DEres[order(DEres$padj), ] # order rows by padj
  DEres <- subset(DEres, padj<0.05)
  return(DEres)
}

create_dir("./DESeq_results_table_padj05")
DEres_csv_outpath <- "./DESeq_results_table_padj05"
```

### Avr/R vs controls

#### i. Gabo_Sr50_AvrSr50 vs Gabo_AvrSr50

```{r}
res <- DEanalysis_results(dds_AllGabo, "Gabo_Sr50_AvrSr50", "Gabo_AvrSr50")
write.csv(as.data.frame(res), file=paste(DEres_csv_outpath, "Gabo_Sr50_AvrSr50_vs_Gabo_AvrSr50_padj05.csv", sep="/"))
res
```

#### ii. Gabo_Sr50_AvrSr50 vs Gabo_EV

```{r echo=FALSE}
res <- DEanalysis_results(dds_AllGabo, "Gabo_Sr50_AvrSr50", "Gabo_EV")
write.csv(as.data.frame(res), file=paste(DEres_csv_outpath, "Gabo_Sr50_AvrSr50_vs_Gabo_EV_padj05.csv", sep="/"))
res
```

#### iii. Gabo_Sr50_AvrSr50 vs Gabo_Sr50_AvrSr35

```{r echo=FALSE}
res <- DEanalysis_results(dds_AllGabo, "Gabo_Sr50_AvrSr50", "Gabo_Sr50_AvrSr35")
write.csv(as.data.frame(res), file=paste(DEres_csv_outpath, "Gabo_Sr50_AvrSr50_vs_Gabo_Sr50_AvrSr35_padj05.csv", sep="/"))
res
```

#### iv. Gabo_Sr50_AvrSr50 vs Gabo_Sr50_EV

```{r echo=FALSE}
res <- DEanalysis_results(dds_AllGabo, "Gabo_Sr50_AvrSr50", "Gabo_Sr50_EV")
write.csv(as.data.frame(res), file=paste(DEres_csv_outpath, "Gabo_Sr50_AvrSr50_vs_Gabo_Sr50_EV_padj05.csv", sep="/"))
res
```

------------------------------------------------------------------------

### GaboSr50 vs Gabo

Comparing Gabo_Sr50 vs Gabo to remove background in Gabo_Sr50 genotype.

#### i) Gabo_Sr50_EV vs Gabo_EV

```{r}
res <- DEanalysis_results(dds_AllGabo, "Gabo_Sr50_EV", "Gabo_EV")
write.csv(as.data.frame(res), file=paste(DEres_csv_outpath, "Gabo_Sr50_EV_vs_Gabo_EV_padj05.csv", sep="/"))
res
```

#### ii) Gabo_Sr50_EV vs Gabo_AvrSr50

```{r echo=FALSE}
res <- DEanalysis_results(dds_AllGabo, "Gabo_Sr50_EV", "Gabo_AvrSr50")
write.csv(as.data.frame(res), file=paste(DEres_csv_outpath, "Gabo_Sr50_EV_vs_Gabo_AvrSr50_padj05.csv", sep="/"))
res
```

#### iii) Gabo_Sr50_AvrSr35 vs Gabo_EV

```{r echo=FALSE}
res <- DEanalysis_results(dds_AllGabo, "Gabo_Sr50_AvrSr35", "Gabo_EV")
write.csv(as.data.frame(res), file=paste(DEres_csv_outpath, "Gabo_Sr50_AvrSr35_vs_Gabo_EV_padj05.csv", sep="/"))
res
```

#### iv) Gabo_Sr50_AvrSr35 vs Gabo_AvrSr50

```{r echo=FALSE}
res <- DEanalysis_results(dds_AllGabo, "Gabo_Sr50_AvrSr35", "Gabo_AvrSr50")
write.csv(as.data.frame(res), file=paste(DEres_csv_outpath, "Gabo_Sr50_AvrSr35_vs_Gabo_AvrSr50_padj05.csv", sep="/"))
res
```

## 4. Extract commonly upregulated genes in Gabo_Sr50_AvrSr50

Rationale:

1.  Filter to retain only upregulated genes from all four Gabo_Sr50_AvrSr50 comparisons (log2FoldChange > 2) *(Note: I have inspected the genes ranked in log2FoldChange order in all dataframes. Using a threshold of 2 seems ok.)*
2.  Intersect and extract upregulated genes common to all comparisons --> Avr/R upregulated gene pool
3.  Repeat step 1 and 2 for control vs control comparisons, and get a list of upregulated genes in Gabo_Sr50 genotype (background)
4.  Remove background from Avr/R upregulated gene pool

Setting up functions

```{r}
upreg_filter <- function(df, lfc) {
  print(paste("Number of rows BEFORE filtering:", dim(df)[1]))
  filt <- df %>% 
    filter(log2FoldChange > lfc) %>% 
    `colnames<-`(c("gene","baseMean","log2FoldChange",
                   "lfcSE","stat","pvalue","padj")) %>% 
    arrange(padj)
  filt_df <- as.data.frame(filt)
  print(paste("Number of rows AFTER filtering:", dim(filt_df)[1]))
  return(filt_df)
  
}

csv2df <- function(comparison) {
  fn <- paste(DEres_csv_outpath, comparison, sep="/")
  fn <- paste(fn, "_padj05.csv", sep="")
  df <- read_csv(file=fn, col_types=cols())
  return(df)
}
```

```{r message=FALSE}
Gabo_Sr50_AvrSr50_vs_Gabo_AvrSr50 <- upreg_filter(csv2df("Gabo_Sr50_AvrSr50_vs_Gabo_AvrSr50"), 2)
Gabo_Sr50_AvrSr50_vs_Gabo_EV <- upreg_filter(csv2df("Gabo_Sr50_AvrSr50_vs_Gabo_EV"), 2)
Gabo_Sr50_AvrSr50_vs_Gabo_Sr50_AvrSr35 <- upreg_filter(csv2df("Gabo_Sr50_AvrSr50_vs_Gabo_Sr50_AvrSr35"), 2)
Gabo_Sr50_AvrSr50_vs_Gabo_Sr50_EV <- upreg_filter(csv2df("Gabo_Sr50_AvrSr50_vs_Gabo_Sr50_EV"), 2)
```

```{r}
df_list <- list(Gabo_Sr50_AvrSr50_vs_Gabo_AvrSr50$gene, 
                Gabo_Sr50_AvrSr50_vs_Gabo_EV$gene, 
                Gabo_Sr50_AvrSr50_vs_Gabo_Sr50_AvrSr35$gene,
                Gabo_Sr50_AvrSr50_vs_Gabo_Sr50_EV$gene)
Gabo_AvrRUpregGenePool <- Reduce(intersect, df_list)
length(Gabo_AvrRUpregGenePool)
```

Do the same for controls for background removal:

```{r warning=FALSE}
Gabo_Sr50_EV_vs_Gabo_EV <- upreg_filter(csv2df("Gabo_Sr50_EV_vs_Gabo_EV"), 2)
Gabo_Sr50_EV_vs_Gabo_AvrSr50 <- upreg_filter(csv2df("Gabo_Sr50_EV_vs_Gabo_AvrSr50"), 2)
Gabo_Sr50_AvrSr35_vs_Gabo_EV <- upreg_filter(csv2df("Gabo_Sr50_AvrSr35_vs_Gabo_EV"), 2)
Gabo_Sr50_AvrSr35_vs_Gabo_AvrSr50 <- upreg_filter(csv2df("Gabo_Sr50_AvrSr35_vs_Gabo_AvrSr50"), 2)

control_df_list <- list(Gabo_Sr50_EV_vs_Gabo_EV$gene, 
                Gabo_Sr50_EV_vs_Gabo_AvrSr50$gene, 
                Gabo_Sr50_AvrSr35_vs_Gabo_EV$gene,
                Gabo_Sr50_AvrSr35_vs_Gabo_AvrSr50$gene)
GaboSr50ControlUpregGenePool <- Reduce(intersect, control_df_list)
length(GaboSr50ControlUpregGenePool)
```

Remove background in Gabo_Sr50_AvrSr50 upregulated gene pool:

```{r}
Gabo_AvrRUpregGenePool <- Gabo_AvrRUpregGenePool[!(Gabo_AvrRUpregGenePool %in% GaboSr50ControlUpregGenePool)]
length(Gabo_AvrRUpregGenePool)
```

```{r}
Gabo_AvrRUpregGenePool
```

#- - - - - - - - - -

# B. Fielder

## 1. Input

### Transcript abundance files

Slice abundance file list to include only Fielder.

```{r}
abundance_files_Fielder <- abundance_files[1:16]
```

### *Tximport*

Using tximport function to import quantification data.

```{r message=FALSE}
txi_Fielder <- tximport(abundance_files_Fielder, type="kallisto", tx2gene=t2g)
```

Construct a sample table with condition column.

```{r}
samples_Fielder <- data.frame(condition=factor(rep(c("Fielder_AvrSr35",
                                             "Fielder_Sr35_AvrSr35",
                                             "Fielder_Sr35_AvrSr50",
                                             "Fielder_Sr35"),
                                           each=4)))
rownames(samples_Fielder) <- colnames(txi_Fielder$counts)
samples_Fielder
```

### Construct a *DESeqDataSet*

Now with the txi object and sample table, construct a DESeqDataSet.

```{r message=FALSE}
dds_Fielder <- DESeqDataSetFromTximport(txi_Fielder, colData=samples_Fielder, design=~condition)
dds_Fielder
```

Remove genes with zero reads, to keep only rows that have at least \>1 read in total across all genes.

```{r}
dds_Fielder <- dds_Fielder[rowSums(counts(dds_Fielder)) > 1, ]
dds_Fielder
```

## 2. Differential expression analysis

```{r message=FALSE}
dds_Fielder <- DESeq(dds_Fielder)
```

```{r}
dds_Fielder
```

## 3. Result tables

Applying *summary()* to a result table will return:\
\* LFC \> 0 (up) --> upregulated genes\
\* LFC \< 0 (down) --> downregulated genes\
\* outliers\
\* low counts\

Since only one genotype is used, Avr/R can be directly compared to all other controls.

------------------------------------------------------------------------

### Avr/R vs controls

#### i. Fielder_Sr35_AvrSr35 vs Fielder_AvrSr35

```{r}
res <- DEanalysis_results(dds_Fielder, "Fielder_Sr35_AvrSr35", "Fielder_AvrSr35")
write.csv(as.data.frame(res), file=paste(DEres_csv_outpath, "Fielder_Sr35_AvrSr35_vs_Fielder_AvrSr35_padj05.csv", sep="/"))
res
```

#### ii. Fielder_Sr35_AvrSr35 vs Fielder_Sr35_AvrSr50

```{r echo=FALSE}
res <- DEanalysis_results(dds_Fielder, "Fielder_Sr35_AvrSr35", "Fielder_Sr35_AvrSr50")
write.csv(as.data.frame(res), file=paste(DEres_csv_outpath, "Fielder_Sr35_AvrSr35_vs_Fielder_Sr35_AvrSr50_padj05.csv", sep="/"))
res
```

#### iii. Fielder_Sr35_AvrSr35 vs Fielder_Sr35

```{r echo=FALSE}
res <- DEanalysis_results(dds_Fielder, "Fielder_Sr35_AvrSr35", "Fielder_Sr35")
write.csv(as.data.frame(res), file=paste(DEres_csv_outpath, "Fielder_Sr35_AvrSr35_vs_Fielder_Sr35_padj05.csv", sep="/"))
res
```

## 4. Extract commonly upregulated genes in Fielder_Sr35_AvrSr35

1.  Filter to retain only upregulated genes from all three Fielder_Sr35_AvrSr35 comparisons (log2FoldChange > 2)
2.  Intersect and extract upregulated genes common to all comparisons --> Fielder vr/R upregulated gene pool
3.  Intersect Fielder_Sr35_AvrSr35 and Gabo_Sr50_AvrSr50 upregulated gene pools --> AvrRUpGenePool (i.e. polished)
4.  Subset DESeqDataSet with the gene ids in polished gene pool, inspect gene statistics
5.  Pick genes that gave low counts (\<20-50) in controls but exceptionally high in Avr/R sets
6.  Make a loop to print all the gene count plots and screen manually.

```{r warning=FALSE}
Fielder_Sr35_AvrSr35_vs_Fielder_AvrSr35 <- upreg_filter(csv2df("Fielder_Sr35_AvrSr35_vs_Fielder_AvrSr35"), 2)
Fielder_Sr35_AvrSr35_vs_Fielder_Sr35_AvrSr50 <- upreg_filter(csv2df("Fielder_Sr35_AvrSr35_vs_Fielder_Sr35_AvrSr50"), 2)
Fielder_Sr35_AvrSr35_vs_Fielder_Sr35 <- upreg_filter(csv2df("Fielder_Sr35_AvrSr35_vs_Fielder_Sr35"), 2)
```

Intersect and extract upregulated genes common to all three comparisons.

```{r}
df_list <- list(Fielder_Sr35_AvrSr35_vs_Fielder_AvrSr35$gene, 
                Fielder_Sr35_AvrSr35_vs_Fielder_Sr35_AvrSr50$gene, 
                Fielder_Sr35_AvrSr35_vs_Fielder_Sr35$gene)
Fielder_AvrRUpregGenePool <- Reduce(intersect, df_list)
length(Fielder_AvrRUpregGenePool)
```

## 5. AvrRUpGenePool

Intersect Gabo and Fielder gene pools to find commonly upregulated genes shared between them.

```{r}
df_list <- list(Gabo_AvrRUpregGenePool, Fielder_AvrRUpregGenePool)
AvrRUpGenePool <- Reduce(intersect, df_list)
AvrRUpGenePool
```

Lastly, export Gabo and Fielder normalised transcript count plots.

```{r fig.width=11, fig.height=5, eval=FALSE}
create_dir("./AvrRUpGenePool_Plots")

counter <- 1
for (gene in AvrRUpGenePool) {
  png(file = paste("./AvrRUpGenePool_Plots/", counter, "_", gene, "_Gabo.png", sep=""), 
       height=400, width=1000)
  plotCounts(dds_AllGabo, gene=gene, intgroup="condition")
  dev.off()
  png(file = paste("./AvrRUpGenePool_Plots/", counter, "_", gene, "_Fielder.png", sep=""), 
       height=400, width=1000)
  plotCounts(dds_Fielder, gene=gene, intgroup="condition")
  dev.off()
  counter <- counter + 1
}
```
